rust_src/src/data.rs

```diff
@@ -181,6 +183,50 @@ pub fn aref(array: LispObject, idx: EmacsInt) -> LispObject {
     }
 }
 
+/// Store into the element of ARRAY at index IDX the value NEWELT.
+/// Return NEWELT.  ARRAY may be a vector, a string, a char-table or a
+/// bool-vector.  IDX starts at 0.
+#[lisp_fn]
+pub fn aset(array: LispObject, idx: EmacsInt, newelt: LispObject) -> LispObject {
+    if let Some(vl) = array.as_vectorlike() {
+        if let Some(v) = vl.as_vector() {
+            unsafe { CHECK_IMPURE(array.to_raw(), array.get_untaggedptr()) };
+            if idx < 0 || idx >= v.len() as EmacsInt {
+                args_out_of_range!(array, LispObject::from(idx));
+            }
+
+            unsafe { v.set_unchecked(idx as ptrdiff_t, newelt) };
+        } else if let Some(bv) = vl.as_bool_vector() {
+            if idx < 0 || idx >= bv.len() as EmacsInt {
+                args_out_of_range!(array, LispObject::from(idx));
+            }
+
+            unsafe { bv.set_unchecked(idx as usize, newelt.is_not_nil()) };
+        } else if let Some(_tbl) = vl.as_char_table() {
+            verify_lisp_type!(idx, Qcharacterp);
+            unsafe { CHAR_TABLE_SET(array.to_raw(), idx as c_int, newelt.to_raw()) };
+        } else if let Some(record) = vl.as_record() {
+            if idx < 0 || idx >= vl.pseudovector_size() {
+                args_out_of_range!(array, LispObject::from(idx));
+            }
+
+            unsafe { record.set_unchecked(idx as isize, newelt) };
```

Comment:
This handles vectors directly but for other types we still call `aset_original`. Are we able to simplify aset_original with this change? I would have expected some C code to be removed.

Reply:
This is a crutch currently while I port the branches of the if. Yes, I could cut down the C code. Hmmm, maybe I will replace ported branches with asserts.

Reply:
The plan is to remove `aset_original` completely once this PR is complete. In part I am doing this work in public so new contributors can see the process we use when working on a more complex function.